src/Text/Parsing/Parser/String.purs

```diff
@@ -42,140 +42,154 @@ module Text.Parsing.Parser.String
+char :: forall m. Char -> ParserT String m Char
 char c = satisfy (_ == c) <?> show c
 
 -- | Match a `String` exactly *N* characters long.
-takeN :: forall m. Monad m => Int -> ParserT String m String
-takeN n = do
-  ParseState input position _ <- get
+takeN :: forall m. Int -> ParserT String m String
+takeN n = join $ state \state1@(ParseState input position _) -> do
   let { before, after } = splitAt n input
   if length before == n then do
-    put $ ParseState after (updatePosString position before) true
-    pure before
-  else fail ("Could not take " <> show n <> " characters")
+    Tuple (pure before) (ParseState after (updatePosString position before) true)
+  else
+    Tuple (fail ("Could not take " <> show n <> " characters")) state1
 
 -- | Match zero or more whitespace characters satisfying
 -- | `Data.CodePoint.Unicode.isSpace`. Always succeeds.
-whiteSpace :: forall m. Monad m => ParserT String m String
+whiteSpace :: forall m. ParserT String m String
 whiteSpace = fst <$> match skipSpaces
 
 -- | Skip whitespace characters and throw them away. Always succeeds.
-skipSpaces :: forall m. Monad m => ParserT String m Unit
-skipSpaces = skipMany (satisfyCodePoint isSpace)
+skipSpaces :: forall m. ParserT String m Unit
+skipSpaces = ParserT
+  ( mkFn5 \(ParseState input pos _) _ _ _ done -> do
+      let head = takeWhile isSpace input
+      let tail = SCU.drop (SCU.length head) input
+      runFn2 done (ParseState tail (updatePosString pos head) true) unit
+  )
 
 -- | Match one of the BMP `Char`s in the array.
-oneOf :: forall m. Monad m => Array Char -> ParserT String m Char
+oneOf :: forall m. Array Char -> ParserT String m Char
 oneOf ss = satisfy (flip elem ss) <~?> \_ -> "one of " <> show ss
 
 -- | Match any BMP `Char` not in the array.
-noneOf :: forall m. Monad m => Array Char -> ParserT String m Char
+noneOf :: forall m. Array Char -> ParserT String m Char
 noneOf ss = satisfy (flip notElem ss) <~?> \_ -> "none of " <> show ss
 
 -- | Match one of the Unicode characters in the array.
-oneOfCodePoints :: forall m. Monad m => Array CodePoint -> ParserT String m CodePoint
+oneOfCodePoints :: forall m. Array CodePoint -> ParserT String m CodePoint
 oneOfCodePoints ss = satisfyCodePoint (flip elem ss) <~?> \_ -> "one of " <> show (singleton <$> ss)
 
 -- | Match any Unicode character not in the array.
-noneOfCodePoints :: forall m. Monad m => Array CodePoint -> ParserT String m CodePoint
+noneOfCodePoints :: forall m. Array CodePoint -> ParserT String m CodePoint
 noneOfCodePoints ss = satisfyCodePoint (flip notElem ss) <~?> \_ -> "none of " <> show (singleton <$> ss)
 
 -- | Updates a `Position` by adding the columns and lines in `String`.
 updatePosString :: Position -> String -> Position
-updatePosString pos str = case uncons str of
-  Nothing -> pos
-  Just { head, tail } -> updatePosString (updatePosSingle pos head) tail -- tail recursive
+updatePosString = go 0
+  where
+  go ix pos str = case codePointAt ix str of
```

Comment:
Ah this might be a problem. `codePointAt` is linear in `ix`. So then `updatePosString` will be quadratic in `length str`.

Reply:
I do think this routine in general could use a bit of work, however. I believe right now it will consider `\r\n` as two lines since it only ever processes on codepoint at a time.